File: frontend/types/finance.ts
export type FinanceSummary = {
  total_cod_due: number
  total_cod_collected: number
  total_cod_pending: number
  total_payments_received: number
  outstanding_balance: number
}

export type FinanceLedgerShipment = {
  id: number
  tracking_number: string | null
  cod_amount: number
  status: string
  created_at: string | null
  delivered_at: string | null
}

export type FinanceLedgerPayment = {
  id: number
  shipment_id: number | null
  cod_id: number | null
  amount: number
  currency: string
  payment_method: string
  payment_status: string
  transaction_reference: string
  paid_at: string | null
  created_at: string | null
  notes: string | null
}

export type FinanceCustomerLedger = {
  customer_id: number
  customer_name: string | null
  total_cod_due: number
  total_payments: number
  outstanding_balance: number
  shipments: FinanceLedgerShipment[]
  payments: FinanceLedgerPayment[]
}

export type FinanceSettlementItem = {
  shipment_id: number
  cod_amount: number
  collected: boolean
  collected_at: string | null
  customer_name: string | null
}

export type FinanceCourierSettlement = {
  driver_id: number
  driver_name: string | null
  total_collected: number
  total_pending: number
  settlements: FinanceSettlementItem[]
}

export type FinanceHistoryItem = {
  id: number
  type: 'payment' | 'cod'
  amount: number
  reference: string
  created_at: string | null
  notes: string | null
}

export type FinanceHistoryResponse = {
  items: FinanceHistoryItem[]
}

export type FinanceReportResponse = {
  summary: FinanceSummary
  history: FinanceHistoryResponse
  generated_at: string
}

File: backend/app/schemas/finance.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class FinanceSummaryResponse(BaseModel):
    total_cod_due: float
    total_cod_collected: float
    total_cod_pending: float
    total_payments_received: float
    outstanding_balance: float


class CustomerLedgerShipmentItem(BaseModel):
    id: int
    tracking_number: str | None
    cod_amount: float
    status: str
    created_at: datetime | None
    delivered_at: datetime | None


class CustomerLedgerPaymentItem(BaseModel):
    id: int
    shipment_id: int | None
    cod_id: int | None
    amount: float
    currency: str
    payment_method: str
    payment_status: str
    transaction_reference: str
    paid_at: datetime | None
    created_at: datetime
    notes: str | None


class CustomerLedgerResponse(BaseModel):
    customer_id: int
    customer_name: str | None
    total_cod_due: float
    total_payments: float
    outstanding_balance: float
    shipments: list[CustomerLedgerShipmentItem]
    payments: list[CustomerLedgerPaymentItem]


class CourierSettlementItem(BaseModel):
    shipment_id: int
    cod_amount: float
    collected: bool
    collected_at: datetime | None
    customer_name: str | None


class CourierSettlementResponse(BaseModel):
    driver_id: int
    driver_name: str | None
    total_collected: float
    total_pending: float
    settlements: list[CourierSettlementItem]


class CodCollectionResponse(BaseModel):
    success: bool
    cod_id: int
    payment_id: int
    collected_at: datetime | None

    model_config = ConfigDict(from_attributes=True)
